File: app/homekit_payload.py
"""HomeKit setup URI (X-HM://) — aligned with SimonGolms/homekit-code."""

from __future__ import annotations

import re

# https://github.com/SimonGolms/homekit-code/blob/master/src/config/categories.ts
HOMEKIT_CATEGORIES: dict[str, int] = {
    "airConditioner": 21,
    "airport": 27,
    "airPurifier": 19,
    "appleTv": 24,
    "bridge": 2,
    "dehumidifier": 23,
    "door": 12,
    "doorLock": 6,
    "fan": 3,
    "faucet": 29,
    "garage": 4,
    "heater": 20,
    "humidifier": 22,
    "ipCamera": 17,
    "lightbulb": 5,
    "other": 1,
    "outlet": 7,
    "programmableSwitch": 15,
    "rangeExtender": 16,
    "securitySystem": 11,
    "sensor": 10,
    "showerHead": 30,
    "speaker": 26,
    "sprinkler": 28,
    "switch": 8,
    "targetController": 32,
    "television": 31,
    "thermostat": 9,
    "videoDoorBell": 18,
    "window": 13,
    "windowCovering": 14,
}

DEFAULT_HOMEKIT_FLAG = 2  # IP


def pairing_digits(value: str) -> str:
    digits = re.sub(r"\D", "", value or "")
    return digits if len(digits) == 8 else ""


def normalize_setup_id(value: str) -> str:
    s = re.sub(r"[^0-9A-Za-z]", "", (value or "").strip()).upper()
    return s[:4] if len(s) == 4 else ""


def normalize_uri_body(body: str) -> str:
    return re.sub(r"[^0-9A-Za-z]", "", (body or "")).upper()


def to_base36_upper(n: int, width: int = 9) -> str:
    alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    if n <= 0:
        return "0".zfill(width)
    out: list[str] = []
    while n:
        n, rem = divmod(n, 36)
        out.append(alphabet[rem])
    return "".join(reversed(out)).zfill(width)


def compose_setup_uri(
    *,
    category_id: int,
    flag: int = DEFAULT_HOMEKIT_FLAG,
    password: str,
    setup_id: str = "",
    version: int = 0,
    reserved: int = 0,
) -> str:
    """Build ``X-HM://{base36}{setupId}`` (homekit-code composeSetupUri)."""
    payload = version & 0x7
    payload = ((payload << 4) | (reserved & 0xF)) & 0xFFFFFFFF
    payload = ((payload << 8) | (category_id & 0xFF)) & 0xFFFFFFFF
    payload = ((payload << 4) | (flag & 0xF)) & 0xFFFFFFFF
    payload = (int(payload) << 27) | (int(password) & 0x7FFFFFFF)
    base36 = to_base36_upper(int(payload), 9)
    sid = normalize_setup_id(setup_id)
    return f"X-HM://{base36}{sid}"


def category_id_for(name: str) -> int:
    key = (name or "other").strip()
    return HOMEKIT_CATEGORIES.get(key, HOMEKIT_CATEGORIES["other"])


def category_name_for_id(category_id: int) -> str:
    for name, cid in HOMEKIT_CATEGORIES.items():
        if cid == category_id:
            return name
    return "other"


def decode_payload_from_base36(base36: str) -> dict[str, int]:
    try:
        n = int(base36, 36)
    except ValueError:
        return {}
    password = n & 0x7FFFFFFF
    rest = n >> 27
    flag = rest & 0xF
    rest >>= 4
    category_id = rest & 0xFF
    return {"password": password, "flag": flag, "category_id": category_id}


def parse_setup_uri(uri: str) -> dict[str, str] | None:
    """Parse X-HM URI; preserve full alphanumeric body (extended hub payloads)."""
    s = (uri or "").strip()
    if not s.upper().startswith("X-HM://"):
        return None
    body = normalize_uri_body(s[7:])
    if len(body) < 9:
        return None
    base36 = body[:9]
    setup_id = ""
    if len(body) >= 13:
        setup_id = normalize_setup_id(body[-4:])
    elif len(body) > 9:
        setup_id = normalize_setup_id(body[9:])
    return {"base36": base36, "setup_id": setup_id, "uri": f"X-HM://{body}"}


def decode_pairing_from_uri(uri: str) -> str:
    parsed = parse_setup_uri(uri)
    if not parsed:
        return ""
    fields = decode_payload_from_base36(parsed["base36"])
    if fields:
        password = fields["password"]
        digits = str(password)
        return digits.zfill(8) if len(digits) <= 8 else ""
    try:
        n = int(parsed["base36"], 36)
    except ValueError:
        return ""
    password = n & 0x7FFFFFFF
    digits = str(password)
    return digits.zfill(8) if len(digits) <= 8 else ""


def decode_fields_from_uri(uri: str) -> dict[str, str | int]:
    parsed = parse_setup_uri(uri)
    if not parsed:
        return {}
    fields = decode_payload_from_base36(parsed["base36"])
    out: dict[str, str | int] = {"setup_id": parsed["setup_id"]}
    if fields:
        out["homekit_flag"] = int(fields["flag"])
        out["homekit_category"] = category_name_for_id(int(fields["category_id"]))
    return out


def qr_encode_payload(qr_payload: str, manual_code: str = "") -> str | None:
    qr = (qr_payload or "").strip()
    if qr.upper().startswith("X-HM://"):
        parsed = parse_setup_uri(qr)
        return parsed["uri"] if parsed else qr
    digits = pairing_digits(manual_code)
    if digits:
        return None
    return None


def normalize_fields(
    manual_code: str,
    qr_payload: str,
    *,
    homekit_category: str = "other",
    homekit_flag: int = DEFAULT_HOMEKIT_FLAG,
    setup_id: str = "",
) -> dict[str, str | int]:
    """Normalize HomeKit vault fields; derive category/setup ID from URI when present."""
    qr = (qr_payload or "").strip()
    parsed = parse_setup_uri(qr) if qr else None
    digits = pairing_digits(manual_code)
    sid = normalize_setup_id(setup_id)
    if parsed and not digits:
        digits = decode_pairing_from_uri(parsed["uri"])
    if parsed:
        decoded = decode_fields_from_uri(parsed["uri"])
        return {
            "manual_code": digits,
            "qr_payload": parsed["uri"],
            "setup_id": str(decoded.get("setup_id") or sid),
            "homekit_category": str(
                decoded.get("homekit_category") or homekit_category
            ),
            "homekit_flag": int(decoded.get("homekit_flag") or homekit_flag),
        }
    if len(digits) == 8:
        cat_id = category_id_for(homekit_category)
        uri = compose_setup_uri(
            category_id=cat_id,
            flag=int(homekit_flag),
            password=digits,
            setup_id=sid,
        )
        return {
            "manual_code": digits,
            "qr_payload": uri,
            "setup_id": sid,
            "homekit_category": homekit_category,
            "homekit_flag": int(homekit_flag),
        }
    return {
        "manual_code": digits,
        "qr_payload": qr,
        "setup_id": sid,
        "homekit_category": homekit_category,
        "homekit_flag": int(homekit_flag),
    }


def has_scannable_qr(qr_payload: str) -> bool:
    return parse_setup_uri(qr_payload or "") is not None

File: app/static/brand/js/homekit-payload.js
/**
 * HomeKit setup URI (X-HM://) — parity with SimonGolms/homekit-code.
 */
(function (global) {
  const CATEGORIES = {
    airConditioner: 21,
    airport: 27,
    airPurifier: 19,
    appleTv: 24,
    bridge: 2,
    dehumidifier: 23,
    door: 12,
    doorLock: 6,
    fan: 3,
    faucet: 29,
    garage: 4,
    heater: 20,
    humidifier: 22,
    ipCamera: 17,
    lightbulb: 5,
    other: 1,
    outlet: 7,
    programmableSwitch: 15,
    rangeExtender: 16,
    securitySystem: 11,
    sensor: 10,
    showerHead: 30,
    speaker: 26,
    sprinkler: 28,
    switch: 8,
    targetController: 32,
    television: 31,
    thermostat: 9,
    videoDoorBell: 18,
    window: 13,
    windowCovering: 14,
  };

  const CATEGORY_KEYS = Object.keys(CATEGORIES).sort((a, b) =>
    a.localeCompare(b)
  );

  const DEFAULT_FLAG = 2;

  function pairingDigits(value) {
    const d = String(value || "").replace(/\D/g, "");
    return d.length === 8 ? d : "";
  }

  function normalizeSetupId(value) {
    const s = String(value || "")
      .replace(/[^0-9A-Za-z]/g, "")
      .toUpperCase();
    return s.length === 4 ? s : "";
  }

  function normalizeUriBody(body) {
    return String(body || "")
      .replace(/[^0-9A-Za-z]/g, "")
      .toUpperCase();
  }

  function toBase36Upper(n, width) {
    const alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    let num = BigInt(n);
    if (num <= 0n) return "0".padStart(width, "0");
    let out = "";
    while (num > 0n) {
      const rem = Number(num % 36n);
      out = alphabet[rem] + out;
      num = num / 36n;
    }
    return out.padStart(width, "0");
  }

  function composeSetupUri({
    categoryId,
    flag = DEFAULT_FLAG,
    password,
    setupId = "",
    version = 0,
    reserved = 0,
  }) {
    let payload = BigInt(version & 0x7);
    payload = (payload << 4n) | BigInt(reserved & 0xf);
    payload = (payload << 8n) | BigInt(categoryId & 0xff);
    payload = (payload << 4n) | BigInt(flag & 0xf);
    payload = (payload << 27n) | BigInt(Number(password) & 0x7fffffff);
    const base36 = toBase36Upper(payload, 9);
    return `X-HM://${base36}${normalizeSetupId(setupId)}`;
  }

  function categoryIdFor(name) {
    return CATEGORIES[name] ?? CATEGORIES.other;
  }

  function categoryNameForId(categoryId) {
    for (const [name, id] of Object.entries(CATEGORIES)) {
      if (id === categoryId) return name;
    }
    return "other";
  }

  function decodePayloadFromBase36(base36) {
    const alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    let n = 0n;
    for (const ch of String(base36 || "").toUpperCase()) {
      const idx = alphabet.indexOf(ch);
      if (idx < 0) return {};
      n = n * 36n + BigInt(idx);
    }
    const password = Number(n & 0x7fffffffn);
    let rest = Number(n >> 27n);
    const flag = rest & 0xf;
    rest >>= 4;
    const categoryId = rest & 0xff;
    return { password, flag, categoryId };
  }

  function parseSetupUri(uri) {
    const s = String(uri || "").trim();
    if (!s.toUpperCase().startsWith("X-HM://")) return null;
    const body = normalizeUriBody(s.slice(7));
    if (body.length < 9) return null;
    const base36 = body.slice(0, 9);
    let setupId = "";
    if (body.length >= 13) setupId = normalizeSetupId(body.slice(-4));
    else if (body.length > 9) setupId = normalizeSetupId(body.slice(9));
    return { base36, setupId, uri: `X-HM://${body}` };
  }

  function decodePairingFromUri(uri) {
    const parsed = parseSetupUri(uri);
    if (!parsed) return "";
    const fields = decodePayloadFromBase36(parsed.base36);
    if (fields.password !== undefined) {
      return String(fields.password).padStart(8, "0").slice(-8);
    }
    const alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    let n = 0n;
    for (const ch of parsed.base36) {
      const idx = alphabet.indexOf(ch);
      if (idx < 0) return "";
      n = n * 36n + BigInt(idx);
    }
    const password = Number(n & 0x7fffffffn);
    return String(password).padStart(8, "0").slice(-8);
  }

  function decodeFieldsFromUri(uri) {
    const parsed = parseSetupUri(uri);
    if (!parsed) return {};
    const fields = decodePayloadFromBase36(parsed.base36);
    const out = { setup_id: parsed.setupId };
    if (fields.categoryId !== undefined) {
      out.homekit_flag = fields.flag;
      out.homekit_category = categoryNameForId(fields.categoryId);
    }
    return out;
  }

  function normalizeFields(manualCode, qrPayload, opts = {}) {
    const category = opts.homekit_category || "other";
    const flag = Number(opts.homekit_flag ?? DEFAULT_FLAG);
    let setupId = normalizeSetupId(opts.setup_id || "");
    let qr = String(qrPayload || "").trim();
    let digits = pairingDigits(manualCode);
    const parsed = qr ? parseSetupUri(qr) : null;
    if (parsed && !digits) digits = decodePairingFromUri(parsed.uri);
    if (parsed) {
      const decoded = decodeFieldsFromUri(parsed.uri);
      return {
        manual_code: digits,
        qr_payload: parsed.uri,
        setup_id: decoded.setup_id || setupId,
        homekit_category: decoded.homekit_category || category,
        homekit_flag: Number(decoded.homekit_flag ?? flag),
      };
    }
    if (digits.length === 8) {
      const uri = composeSetupUri({
        categoryId: categoryIdFor(category),
        flag,
        password: digits,
        setupId,
      });
      return {
        manual_code: digits,
        qr_payload: uri,
        setup_id: setupId,
        homekit_category: category,
        homekit_flag: flag,
      };
    }
    return {
      manual_code: digits,
      qr_payload: qr,
      setup_id: setupId,
      homekit_category: category,
      homekit_flag: flag,
    };
  }

  function hasScannableQr(qrPayload) {
    return parseSetupUri(qrPayload) !== null;
  }

  function codeProtocol(code) {
    const ct = String(code?.code_type || "").toLowerCase();
    if (ct === "homekit") return "homekit";
    if (hasScannableQr(code?.qr_payload)) return "homekit";
    if (String(code?.qr_payload || "").toUpperCase().startsWith("MT:")) return "matter";
    return "matter";
  }

  function formatPairingDisplay(manualCode) {
    const d = pairingDigits(manualCode);
    if (d.length !== 8) return String(manualCode || "").trim();
    return `${d.slice(0, 3)}-${d.slice(3, 5)}-${d.slice(5, 8)}`;
  }

  global.RemattersHomeKitPayload = {
    CATEGORIES,
    CATEGORY_KEYS,
    DEFAULT_FLAG,
    pairingDigits,
    normalizeSetupId,
    normalizeUriBody,
    composeSetupUri,
    categoryIdFor,
    categoryNameForId,
    parseSetupUri,
    decodePairingFromUri,
    decodeFieldsFromUri,
    normalizeFields,
    hasScannableQr,
    codeProtocol,
    formatPairingDisplay,
  };
})(typeof window !== "undefined" ? window : globalThis);
